lock-api/src/test/java/com/palantir/lock/client/LockTokenShareTest.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * (c) Copyright 2019 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.lock.client;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+import java.util.List;
+import java.util.UUID;
+import java.util.stream.Collectors;
+
+import org.junit.Test;
+
+import com.palantir.lock.v2.LockToken;
+
+public class LockTokenShareTest {
+    private static final LockToken LOCK_TOKEN = LockToken.of(UUID.randomUUID());
+
+    @Test
+    public void allSharedLocksHaveCorrectUnderlyingToken() {
+        List<LockTokenShare> tokens = getSharedLockTokens(LOCK_TOKEN, 3);
+        assertThat(tokens)
+                .extracting(LockTokenShare::sharedLockToken)
+                .containsOnly(LOCK_TOKEN);
+    }
+
+    @Test
+    public void allSharedTokensShouldHaveDifferentIds() {
+        List<LockToken> tokens = LockTokenShare.share(LOCK_TOKEN, 3);
+        assertThat(tokens)
+                .extracting(LockToken::getRequestId)
+                .doesNotHaveDuplicates()
+                .hasSize(3);
+    }
+
+    @Test
+    public void shouldReturnUnderlyingTokenAfterAllReferencesAreUnlocked() {
+        List<LockTokenShare> tokens = getSharedLockTokens(LOCK_TOKEN, 2);
+        LockTokenShare firstToken = tokens.get(0);
+        LockTokenShare secondToken = tokens.get(1);
+
+        assertThat(firstToken.unlock()).isEmpty();
+        assertThat(secondToken.unlock()).contains(LOCK_TOKEN);
```

Comment:
reading this test again, I think we also want to have a third assert: (also added `as` descriptions to describe each state/assert, since it's "stateful")

```java
assertThat(firstToken.unlock())
  .as("no token as second token still has reference")
  .isEmpty();
assertThat(secondToken.unlock())
  .as("get lock token since both references are unlocked")
  .contains(LOCK_TOKEN);
assertThat(firstToken.unlock())
  .as("first token now has lock token since both references unlocked")
  .contains(LOCK_TOKEN);
```